### DISCUSSION TITLE
[V3] Snapshot missing on Livewire component with id: " is not a constructor When there is a child component

### DISCUSSION THREAD

**Discussion by romansh** [livewire/livewire#5935] [1. Help]
#app\Livewire\Test.php
```php
<?php

namespace App\Livewire;
use Livewire\Component;

class Test extends Component
{

    public $text;

    public function save()
    {
        $this->validate([ 
            'text' => 'required',
        ]);
    }

    public function render()
    {
        return view('livewire.test');
    }
}
```
#resources\livewire\test.blade.php

```php 
<div>
    <form wire:submit="save">
        <textarea wire:model="text"></textarea>
        @error('text') <div class="font-semibold text-red-600 mt-0 mb-2">{{ $message }}</div> @enderror
        <livewire:test2 />
        <button type="submit">Save</button>          
    </form>
</div>
```

#app\Livewire\Test2.php
```php 
<?php

namespace App\Livewire;

use Livewire\Component;

class Test2 extends Component
{
    public function render()
    {
        return view('livewire.test2');
    }
}
```

#resources\livewire\test2.blade.php
```php 
`<div>
    {{-- In work, do what you enjoy. --}}
</div>`
```

When I click on the Save button I get 

`Uncaught TypeError: "Snapshot missing on Livewire component with id: " is not a constructor
    at new Component (livewire.js?id=3605227a:3798:15)
    at initComponent (livewire.js?id=3605227a:3846:21)
    at livewire.js?id=3605227a:7848:26
    at livewire.js?id=3605227a:1591:58
    at livewire.js?id=3605227a:1189:42
    at Array.forEach (<anonymous>)
    at livewire.js?id=3605227a:1189:27
    at walk (livewire.js?id=3605227a:1112:5)
    at livewire.js?id=3605227a:1187:7
    at deferHandlingDirectives (livewire.js?id=3605227a:1002:5)`
    
  If I take away <livewire:test2 />, the error disappears. I was trying to change the validation rules to Attribute Rules, but that did not help. It seems the issue is too serious for it to be a missed and unfixed bug.. Where can I take a look to understand what I did wrong in the upgrade to v3?

**Comment by edalzell** [comment]
This is a known issue and Josh has written a failing test, so it will be resolved shortly, I'm sure.

https://github.com/livewire/livewire/discussions/5860

**Reply by romansh** [reply]
Hope it's the same issue. The only reason for my doubt is that it does not have an obvious empty div in the parent component. Tomorrow I will look more closely.

**Comment by joshhanley** [comment]
@romansh yep, we will be tagging beta.4 a bit later so that should hopefully fix this issue. Can you test it once it has been released and let me know how it goes? Thanks!

**Comment by joshhanley** [comment]
@romansh just did a test with your supplied code (thanks for that makes my job easier!) and it seems to be fixed now with the latest changes. At the moment, seeing as we're in beta, no issue is "too serious for it to be a missed and unfixed bug" 😅

Your issue was related to the `@error` tag about it which is sometimes present and sometimes not.

So once I tag beta.4 your issue should be solved 🙂

**Reply by romansh** [reply]
The issue has not gone away with beta4 :(

>Your issue was related to the @error tag about it which is sometimes present and sometimes not.

When I wrap the line with @error in the Blade template as shown below, the Livewirel component behaves as expected.
```blade
@if($errors->any())
            @error('text') <div class="font-semibold text-red-600 mt-0 mb-2">{{ $message }}</div> @enderror
 @endif
```

**Reply by romansh** [reply]
With $errors->any()

```
<div wire:id="apNHXyfJa3oqOYUtoiHd">
    <form wire:submit="save">
        <input type="text" wire:model="text">
        <!-- __BLOCK__ -->             <div class="font-semibold text-red-600 mt-0 mb-2">The text field is required.</div>          <!-- __ENDBLOCK__ -->
        <div wire:snapshot="{&quot;data&quot;:[],&quot;memo&quot;:{&quot;id&quot;:&quot;uMRv2QZGZfPn9DIR0GHG&quot;,&quot;name&quot;:&quot;test2&quot;,&quot;path&quot;:&quot;test&quot;,&quot;method&quot;:&quot;GET&quot;,&quot;children&quot;:[],&quot;lazyLoaded&quot;:true,&quot;errors&quot;:[],&quot;locale&quot;:&quot;en&quot;},&quot;checksum&quot;:&quot;914cc72fae5a071c816f6e90c4401336ea8ce2bade3f0532dc1feacbe3565cf2&quot;}" wire:effects="[]" wire:id="uMRv2QZGZfPn9DIR0GHG">
    
</div>        <button type="submit">Save</button>          
    </form>
</div>
```

Without $errors->any()

```
<div wire:id="WpxwspVNZ6XweNGbXM56">
    <form wire:submit="save">
        <input type="text" wire:model="text">
         <div class="font-semibold text-red-600 mt-0 mb-2">The text field is required.</div>         <div wire:id="RNbLpWkpd1UmiB4ZttjN"></div>        <button type="submit">Save</button>          
    </form>
</div>

```

**Reply by romansh** [reply]
If I add a wire.loading block: 

```
<div>
    <form wire:submit="save">
        <textarea wire:model="text"></textarea>
        @error('text') <div class="font-semibold text-red-600 mt-0 mb-2">{{ $message }}</div> @enderror
        <livewire:test2 />
        <div class="mt-2 flex justify-end">
            <button class="bg-gray-800 text-white rounded p-2" type="submit">      
                Save
            </button>
            <div wire:loading> 
                {{ __('Saving post...') }}
            </div>
        </div>         
    </form>
</div>
```
the issue dissapears.

The same effect with:

```
<div x-cloak> 
              1111
            </div>
```

instead of 

```      

<div wire:loading> 
                {{ __('Saving post...') }}
            </div>
```

But if the block is visible (without x-cloak for example) it's not working and the snapshot error is there. Miracles.

**Reply by joshhanley** [reply]
@romansh thanks for testing. I think I know what is causing it, but not 100% on how to fix it, so I'll try and get Caleb to have a look.

**Comment by joshhanley** [comment]
@romansh this should now be fixed in beta.5. Can you test it and let me know if there are still issues? Thanks!

**Comment by N1ebieski** [comment]
It still doesn't work for me (beta.5)

```html
        <form wire:submit.prevent="submit">
            <div>
                <input 
                    type="text" 
                    class="form-control {{ $errors->has('input') ? 'is-invalid' : ($errors->isNotEmpty() ? 'is-valid' : '') }}" 
                    wire:model="input" 
                >
                @error('input') 
                <span class="invalid-feedback">{{ $message }}</span>
                @enderror
            </div>
            <button>
                Submit
            </button>
        </form>
```

[screen-recording(1).webm](https://github.com/livewire/livewire/assets/14810714/81b6b388-28a2-4a3a-bd98-f371a8e1d07b)

**Comment by romansh** [comment]
Yep, can still see the issue in beta.5.

But what I have noticed: 

```php
<div>
    <form wire:submit="save">
        <textarea wire:model="text"></textarea>
        @error('text'){{ $message }}@enderror
        <livewire:test2 />
        <button type="submit">Save</button>          
    </form>
</div>
```
When I remove the div around the error message the 'snapshot missinng' does not show up.

update: 
without showing an error also when the div outside the @error/@enderror block.

```php
<div>
    <form wire:submit="save">
        <textarea wire:model="text"></textarea>
        <div class="font-semibold text-red-600 mt-0 mb-2">
            @error('text'){{ $message }}@enderror
        </div>
        <livewire:test2 />
        <button type="submit">Save</button>          
    </form>
</div>
```

**Reply by joshhanley** [reply]
@romansh thanks for testing! Any chance you could submit a failing test PR so we can investigate it further? Thanks!

**Comment by lonnylot** [comment]
I am having this issue on beta6 as well. Though, I'm not using `@error`. Just looping over a list of conversations to show them in an <li>

**Reply by lonnylot** [reply]
Posted a failing test: https://github.com/livewire/livewire/pull/6166

**Comment by jigar-dhulla** [comment]
In my case, child component had a form. and it was nested inside a form of parent component. 

Made sure child component was not under form and error went away.

**Comment by romansh** [comment]
@joshhanley, with beta.7, my issue has gone. I tried to write a failing test but could not reproduce the issue during testing. I tried this and that; nonetheless, it was not showing any errors. Then I tried to update composer in 'real' project, and the issue, as I wrote above, has gone.

**Reply by romansh** [reply]
"But now it writes in the console such text "(el2) => {
        if (isntElement(el2))
          return;
        return el2.hasAttribute(`wire:key`) ? el2.getAttribute(`wire:key`) : el2.hasAttribute(`wire:id`) ? el2.getAttribute(`wire:id`) : el2…""

**Reply by joshhanley** [reply]
@romansh yeah there was a stray `console.log` left behind, shouldn't impact how it runs though. It has been cleaned up 🙂

**Comment by N1ebieski** [comment]
For me beta.7 still doesn't work properly. But my issue is similar to #6030, an error "Snapshot missing on Livewire component with id: " is not a constructor" appears when I click the back button in the browser

**Reply by joshhanley** [reply]
@N1ebieski thanks! Can you comment on that issue with your details and code, if the issue is similar? Thanks!

**Comment by nam-co** [comment]
Hello, Just in case is related, in this code, I'm always getting the error bellow

```HTML
//on slider change, it changes the alpine value, but onces $wire.set updates the php variable, the alpine mt_value is set reverted
//This included multiple times on the page using @include
<figure x-data="{mt_value: {{$margin_top}} }">
        <input type="range" min="0" max="40" step="1" x-model="mt_value" x-on:change="$wire.set('itemsArray.{{$itemKey}}.margin_top', $event.target.value)">
</figure>
```

```JS
(el2) => {
        if (isntElement(el2))
          return;
        return el2.hasAttribute(`wire:key`) ? el2.getAttribute(`wire:key`) : el2.hasAttribute(`wire:id`) ? el2.getAttribute(`wire:id`) : el2.id;
      }
TypeError: window.Alpine.cloneNode is not a function. (In 'window.Alpine.cloneNode(from2, to)', 'window.Alpine.cloneNode' is undefined)
```

Sorry if it's not related

**Reply by calebporzio** [reply]
You're alpine assets are out of date (unrelated to the above issues)

In Livewire 3, Alpine is included with Livewire so make sure you remove any alpinejs cdns you had loaded before and that should be fixed.

**Reply by nam-co** [reply]
Thanks for the response Caleb, I double check and is just the @livewireScripts , the strange thing is that this just started on B7, B5 was ok (I think I skip B6)

**Reply by calebporzio** [reply]
Yeah the reason it started is I added a new method to Alpine JUST for beta.7 called "cloneNode"

So if you are somehow using an outside version of Alpine you would get this error

**Reply by MrFabri** [reply]
Removing @livewireScripts and @livewireStyles from blade layouts works for me.
[Livewire V3 Installation](https://livewire.laravel.com/docs/installation)

And removing a empty `<div>` inside `<tr> `

**Reply by nam-co** [reply]
@calebporzio hope you'll consider soon wire:if wire:foreach..forelse etc, or maybe @wireIf, also a first party of x-autoanimate 🤞. I'm closing the issue, thanks for the excellent product

**Comment by fabiopacifici** [comment]
Hi Guys, is this a livewire bug or it has something to do with how the code is written? 

I am having the same error message when i nest a search box component inside a full page component.  I want to create a dynamic search box that filters down results in a table below.

The following behaviour seems to work fine:
- I type in the searchbox a word that can be found in the posts
- the query property is updated and an event is dispatched
- the searchPosts method is called in the full page component
- the posts are filtered and only a bounch are retrived
-  [the error is in console anyway]

This flow won't work:
The search box has text of the previous search
 when I delete some characters from the search box

 Expected:
 -posts are filtered again returning only matching results
 
Actual:
- the error message above is in console 
- nothing apparently happens
- the snapshot missing error is in console (from the previous search)
- a new error pops up in console `livewire.esm.js:7901 Uncaught (in promise) undefined`
- posts are not filtered back


This other flow instead works:
- i type in the search box some random text `sdfsdfsadgsgfd`
- posts are filtered and no results returned
- when the entire string is deleted from the search 
- posts are filtered back and they all appear in the table

```output
Uncaught TypeError: "Snapshot missing on Livewire component with id: " is not a constructor
    at new Component (livewire.esm.js:8469:13)
    at initComponent (livewire.esm.js:8517:19)
    at livewire.esm.js:8736:24
    at livewire.esm.js:2449:60
    at livewire.esm.js:2047:44
    at Array.forEach (<anonymous>)
    at livewire.esm.js:2047:29
    at walk (livewire.esm.js:1970:7)
    at livewire.esm.js:2045:9
    at deferHandlingDirectives (livewire.esm.js:1860:7)
```

This is my search-posts component that i use in the full page PostsPage component.
    
```php
    <livewire:search-posts />
```
    
its code is as follow
```php
class SearchPosts extends Component
{

     #[Rule('string|max:255')]
    public $query;

    function updated($property)
    {
        //dd($property);
        if ($property === 'query') {
            $this->validate();
            //dd($property, $this->query);
            $this->dispatch('searching-posts', $this->query)->to(PostsPage::class);
        }
    }


    public function render()
    {
        return <<<'HTML'
        <div x-data="{showSearch: false}">

            <button class="btn btn-outline" x-on:click="showSearch = !showSearch">
                <i class="bi bi-search"></i>
            </button>


            <div class="w-100 pt-3 pb-4" x-show="showSearch">
                    <input name="query" type="text" class="form-control position-absolute start-0 top-1" aria-label="Button" aria-describedby="" placeholder="Start typing" wire:model.live.debounce.500ms="query" x-transition>
            </div>

        </div>
        HTML;
    }
}
```
When you click on the icon, the search input pops up. When you type, the event is dispatched and posts are initially filtered down but the error message is in console. When deleting characters in the search box, the posts should be filtered again as the input uses the `model.live` modifier.

This is the PostsPage component

```php

class PostsPage extends Component
{
    use WithPagination;


    #[On('searching-posts')]
    public function searchPosts($query = '')
    {
        $this->posts = Post::query()
            ->when($query, function ($queryBuilder) use ($query) {
                $queryBuilder->where('title', 'like', '%' . $query . '%')
                    ->orWhere('content', 'like', '%' . $query . '%');
            })
            ->orderByDesc('updated_at')
            ->paginate(10);
    }

    function mount()
    {
        $this->searchPosts();
    }

    public function render()
    {
        return view('livewire.posts-page', ['posts' => $this->posts])->layout('components.layouts.admin');
    }
}

```

UPDATE:
Asked chat and replied as follow:

From the error message, it seems like Livewire is having trouble finding the component with the given ID. This could be due to a few reasons:

👉Component Nesting: Livewire does not support nesting of components. If you're trying to nest SearchPosts inside PostsPage, it might be causing the issue.

Component Naming: Make sure the component is correctly named and the name matches in all places. Livewire uses the name to find the component, so any discrepancy can cause issues.

Component Registration: Ensure that the component is correctly registered in your AppServiceProvider or wherever you're registering your Livewire components.

Livewire Version: There might be some breaking changes in Livewire v3 that are causing this issue. You might want to check the [Livewire upgrade guide](https://laravel-livewire.com/docs/3.x/upgrade-guide) to see if there's anything that might be causing this.

JavaScript Errors: The error seems to be coming from the JavaScript side of things. Make sure you're not getting any JavaScript errors before this one that might be causing the issue.

Event Dispatching: There might be an issue with the way you're dispatching and handling the event. Make sure the event is correctly dispatched and the listener is set up correctly.

Try to debug these points and see if you can find the issue. If you're still having trouble, please provide more information about your setup and I'll be happy to help further.


any help would be very appreciated! 🙏

**Reply by ojsoft** [reply]
@Component Nesting
Taken from the Original livewire page
"Nesting
Like many other component-based frameworks, Livewire components are nestable — meaning one component can render multiple components within itself."

So this is blank false. I nest components all the time. This is one of the basic features of livewire. You just have to take care that you use wire:key on all nested components.

**Comment by joydeep-bhowmik** [comment]
I'm having the same problem ... everyday new problems with livewire 3 feels like it's still in beta

**Reply by ojsoft** [reply]
I upgraded to livewire 3 a week ago and I could fill a book with problems. Some due to changes that I can follow, just behavior that is different and kind of makes my live hard, but some are really odd, like this one. It is a big mess and I got to far to go back already.

**Reply by ojsoft** [reply]
Well I made the update to get rid of a bug in LiveWire 2. This worked ;-) So life is giving and taking.

**Comment by ojsoft** [comment]
Maybe of any help for someone else:

This code produced a lot of problems in livewire 3 in my case, in livewire 2 it worked like a charm.

```
@foreach($members as $member)
<div class="flex">
<div class="flex-1">
<livewire:member :member="$member" wire:key="member-{{ $member->id }}"/>
</div>
</div>
@endforeach

```
This was the healing, why it is not not enough to key the livewire components itself, I do not know.

```
@foreach($members as $member)
<div class="flex" wire:key="flex-{{ $member->id }}">
<div class="flex-1">
<livewire:member :member="$member" wire:key="member-{{ $member->id }}"/>
</div>
</div>
@endforeach

```

Uhhh, i Figured out that the problem also persists, if there is no livewire component inside the foreach. Just to have wire:model and stuff inside any foreach in a livewires blade. Problems appear as soon as values get changed by model.live calls. Then really weird things happen.

**Reply by shokanshi** [reply]
Pretty much same situation. There is no child component and as long as wire:model.live is in use, weird things start happening. The moment .live is removed, the component will behave properly (except it loses the ability of .live lifecycle).

Anybody knows of why .live is breaking the livewire component in 3.0.5?

EDIT: Another weird situation that I have encounter is form submission using the Enter key and clicking [button type="submit"] seems to have different behavior. Clicking the [button] will tend to prevent child component from full rendering.

**Reply by lrljoe** [reply]
Please read the docs carefully around the behaviour of loops, as it explains clearly that you need to key the first element beneath any foreach loop in a view/blade.

There is nothing "broken" about the behaviour.  If you don't want to migrate to Livewire 3, then you can remain on Livewire 2.  The docs for v2 **advised** to use keys on loops.  v3 just enforces that guidance.  This is necessary to progress the framework.

**Reply by shokanshi** [reply]
The reason for me to switch to v3 is because I am still in the early stages of development and I thought it will be wise for me to go with the latest version instead of sticking to v2.

I did read the document and followed the instructions (Unless I misunderstood how it work). Here is my blade file code of the child component that is throwing the javascript error:

```
@if (!$showResult)
    <div x-data="{
        checkScroll() {
            let load = false;
            window.onscroll = function(ev) {
                if (!load && (window.innerHeight + window.scrollY) >= document.body.scrollHeight) {
                    load = true;
                    @this.call('loadNextPage')
                }
            };
        }
    }" x-init="checkScroll" class="flex justify-center p-2" wire:key="{{ $key }}">
        <svg wire:loading class="animate-spin h-5 w-5" xmlns="http://www.w3.org/2000/svg" fill="none"
            viewBox="0 0 24 24">
            <circle class="opacity-25" cx="12" cy="12" r="10" stroke="currentColor" stroke-width="4">
            </circle>
            <path class="opacity-75" fill="currentColor"
                d="M4 12a8 8 0 018-8V0C5.373 0 0 5.373 0 12h4zm2 5.291A7.962 7.962 0 014 12H0c0 3.042 1.135 5.824 3 7.938l3-2.647z">
            </path>
        </svg>
    </div>
@else
    <div x-data class="min-w-full">
        <table class="relative min-w-full divide-y-2 divide-gray-200 text-sm bg-white">
            <colgroup>
                <col width="">
                <col width="">
            </colgroup>
            <thead>
                <tr class="m-0">
                    <th class="sticky top-0 z-40 px-2 py-2 text-left font-medium text-gray-900 bg-gray-100/95">
                        {{ __('Date') }}
                    </th>
                    <th class="sticky top-0 z-40 px-2 py-2 text-center font-medium text-gray-900 bg-gray-100/95">
                        {{ __('Hrs') }}
                    </th>
                </tr>
            </thead>
            <tbody>
                @foreach ($items as $item)
                    <tr class="m-0" wire:key="{{ $item->id }}">
                        <td class="px-2 py-2 text-gray-700 text-start space-x-2">
                            {{ $item->actual_start }}
                        </td>
                        <td class="px-2 py-2 text-gray-700 text-center space-x-2">
                            {{ $item->approved_hours }}
                        </td>
                    </tr>
                @endforeach
            </tbody>
        </table>

        @if ($items->hasMorePages())
            <livewire:test.list-more-component :page="$this->getPage()" :uuid="$uuid" :showResult="false"
                :key="$key" />
        @endif
    </div>
@endif
```

A brief walkthrough on what the code is doing:

1. Default it will display the results under the `@else` block, at the end of the result, 
2. If there are more pages, it will do a recursive call on itself passing in the current page, and the instruction not to show the result but instead the content inside `@if`
3. When I scroll down the result to the bottom, the javascript will perform a `@this.call('loadNextPage')` to populate the result in this child component and the error will show up in the console.

The code for the child component:

```
    use WithPagination;
    public string $uuid;
    public $showResult = false;
    public $page = 1;
    protected $queryString = [
        'search',
    ];

    public function mount($page = 1)
    {
        $this->setPage($this->page = $page);
    }

    public function render()
    {
        $key = "page-{$this->getPage()}-key";

        if (! $this->showResult) {
            return view('livewire.test.list-more-component')->with([
                'key' => $key,
            ]);
        }

        $items = EmployeeWorkingHour::query()
            ->where('employee_id', $this->uuid)
            ->orderBy('first_name')
            ->orderBy('actual_start')
            ->paginate(50, ['*'], null, $this->getPage());

        return view('livewire.test.list-more-component')->with([
            'key' => $key,
            'items' => $items,
        ]);
    }

    public function loadNextPage()
    {
        $this->setPage($this->getPage() + 1);
        $this->showResult = true;
    }
```

The following is a screenshot from devtool before the scroll event:
<img width="613" alt="image" src="https://github.com/livewire/livewire/assets/10794342/6cb9fc77-0412-4dc9-b6c5-9df8f707f3ae">

After the scroll event, the <svg> block containing the snapshot was totally replaced, as shown below:
<img width="663" alt="image" src="https://github.com/livewire/livewire/assets/10794342/216e2bad-2cf4-4fa6-bc44-02455bd3d3b8">

My understanding is that the replaced block should also contain the snapshot but apparently, after replacement, the snapshot was not added back.

**Reply by lrljoe** [reply]
Add a 
```
<div> 
```
at the top of your blade, and close it at the end.

Don't add any attributes to it!

As I advised earlier, adding wire attributes to root attributes can cause the snapshot to not be maintained 

Where you're using an id as a key, prefix it with something, otherwise you'll end up with two elements with a wire key of "3" for example

e.g.
```
wire:key="something-{{ $item->id }}"
```

Where you're setting the key to page-... again, give it a prefix, otherwise you may end up with key clashes

**Reply by shokanshi** [reply]
Okay enclosing the whole blade within `<div>` at the root fixed the issue for me. Thank you very much! This was driving me crazy because the same piece of code worked in v2.

In the case of `wire:key="something-{{ $item->id }}"`, id is uuid so I did not prefix it with anything. Probably I should make it a habit and just always prefix it.

**Reply by lrljoe** [reply]
No problem, if you could mark the answer and close the discussion, then that'll help other people find the solution!

**Reply by shokanshi** [reply]
I would love to mark my part as answered but I am not the OP of this discussion (I didn't open a new discussion as I didn't want to create multiple threads of the same issue) so I don't think I can (unless I am wrong)...

**Reply by lrljoe** [reply]
Ha, my mistake!

**Comment by f-ricci** [comment]
Had the same problem with the current version of livewire. Apparently V3 requires the root element inside the foreach to have its own unique key.

**Comment by mabdalmoniem** [comment]
I had the same issue, my previous code was rendering the errors like this:
(Which worked like a charm on v2)

```
@if ($errors->any())
    //
@endif    
```

the solution was to wrap it with a div

```
<div>
@if ($errors->any())
    //
@endif    
</div>
```

I dunno what is happening but it's fixed for now

**Reply by ahmedabdel3al** [reply]
just to be carefully , if we put any if condition before any other livewire component we need to wrap it inside div i dunno why 
,but if we putted after you don't have to wrap it 

```
@if ($errors->any())
    //
@endif   
<livewire:component /> 
```
we have to wrap if with div

 ```
<div>
  @if ($errors->any())
      //
  @endif   
</div>
<livewire:component /> 
```
but if we  put @if after we don't have to do wrap it  with div 

 ```
 <livewire:component /> 
  @if ($errors->any())
      //
  @endif   
```

**Comment by iosc** [comment]
hi all, I have the similar error of "missing snapshot" case

after i went through everything here then i still have the issue..

finally i solved it!

my problem is my ```wire:key``` in the wrong place 

The code below having problem
```
@foreach ($receipts as $receipt)
    <div  wire:key="Rcpt-{{ $receipt->id }}">
            ...
            ...
        <div>
            <livewire:item-orders :receipt_id="$receipt->id">
        </div>
    </div>
@endforeach
```

Solved by putting in the correct ```wire:key``` 
```
@foreach ($receipts as $receipt)
    <div>
            ...
            ...
        <div>
            <livewire:item-orders :receipt_id="$receipt->id"  wire:key="io-{{ $receipt->id }}">
        </div>
    </div>
@endforeach
```

**Comment by QubeX2** [comment]
I had this problem with errors too. When I added wire:key it started working again
`@if($errors->any())
    <ul>
        @foreach($errors->all() as $error)
            <li wire:key="{{Str::random(16)}}" class="p-1 text-sm rounded-md border border-red-500 bg-red-200 text-red-900">{{$error}}</li>
        @endforeach
    </ul>
@endif`

**Comment by nvahalik** [comment]
I, too, have had this error on a site we just last week updated to Livewire v3.

What seems to be the issue is having multiple adjacent components. My component look(ed) like this:

    <button class="btn" wire:click="$toggle('addBuddy', true)">Add Buddy</button>

    @if($addBuddy)
        @livewire('user.buddy.add', ['user' => $user])
    @endif

    @livewire('user.buddy.listing', ['user' => $user])


Adding wrapping `div` elements did not help. Moving the second component into an `@else` block seems to have fixed it, but I don't know why!

**Reply by ryanmortier** [reply]
Is your nested component inside a parent component form?

**Reply by nvahalik** [reply]
It's not a form, but it is a Livewire component.

**Reply by N1ebieski** [reply]
Have you tried adding a key as the 3rd parameter?

```php
@livewire('user.buddy.add', ['user' => $user], key('user-buddy-add'))

@livewire('user.buddy.listing', ['user' => $user], key('user-buddy-listing'))
```

**Reply by ryanmortier** [reply]
> Have you tried adding a key as the 3rd parameter?

Yes and this fixed it in my case but I'm unsure why it's needed because the child component is not in a loop.

**Reply by chatisk** [reply]
> > Have you tried adding a key as the 3rd parameter?
> 
> Yes and this fixed it in my case but I'm unsure why it's needed because the child component is not in a loop.

From my experience migrating from Livewire 2 to Livewire 3, wire keys are needed almost everywhere you have a livewire component, whether it's in a loop or not. It helps with the framework keeping track of the components.

**Comment by francoism90** [comment]
Edit: so it turns out I indeed had same model keys..

I was merging collections containing models, but I did fail to notice the it wasn't done correctly, and I ended up with the same models keys.

Best way to find out is by dumping your final collection or query result, as this may lead to issues like I had, and it keeps you up at night lol.

**Reply by ryanmortier** [reply]
You're thinking this issue might have been brought on my a minor update? I can probably test this by installing older versions.

Edit, still exists in 3.5.0 and I'm unable to downgrade to 3.4 because my project uses a dependency with LW ^3.5

**Comment by ojsoft** [comment]
Another pitfall is if elements appear conditionally. Livewire gets confused than and sometimes Evan wire:key does not seem to help.
If I get to a problem like that I make elements only invisible instead of removing them conditionally. This helps always.

**Reply by chatisk** [reply]
I have switched many `@ifs` to x-show for that reason

**Comment by ryanmortier** [comment]
Ah ha, I fixed my issue by following the guidance here: https://livewire.laravel.com/docs/morphing#wrapping-conditionals

Wrapped my conditional in a `<div></div>` and it's fixed.

**Comment by caercam** [comment]
Just spent a couple hours on this one too. It's the middle of the night so I'm not digging any further, but after trying every bit of solution here and there, what did the trick for me was to add random text to the keys. None of these worked:

- `wire:key="export-status"`
- `wire:key="export-{{ $model->id }}"`
- `wire:key="export-status-{{ $model->id }}"`
- `key('export-status')`
- `key("export-status-{{ $model->id }}")`

What did work:

- `wire:key="export-status-{{ $model->id }}-{{ Str::random(16) }}"`

This happens after components are rerender (faulty component is part of a filterable index; everything works when listing models, try to filter the list with Livewire, error ensue). So as @ojsoft said it looks like Livewire is getting confused with itself and his keys. Adding random stuff seems to help. In case it helps anyone --including me if I completely forget about this thing and find this message in couple years. Hello future me!

**Comment by chatisk** [comment]
I have encountered this issue multiple times.
1. First it was fixed by enclosing `@ifs` inside `<div>`
2. Then adding wire:keys="my-dummy-text" (the key is the same, not changing)
3. Adding wire:key={{ Str::uuid }} or anything that changes everytime it re-renders, sometimes makes the component lose its state, so it was not preferrable, but sometimes it doesn't.

Generally, i am a bit confused.

**Reply by francoism90** [reply]
@chatisk It's perfectly valid to be confused, because unfortunately Livewire doesn't handle this automatically. I'm working with Livewire for 3 years now, and I still have these issues sometimes.

Compared to most JS-frameworks they work with a single `key="foo"`, but Livewire has two. You have `wire:key`, which should be used on you HTML-code, and `key` that should be used when nesting Livewire components. In most cases you need both, a `<li wire:key="$i">`, and if the item  inside the loop is a Livewire component, you need a `key` as well:

```blade
@foreach ($items as $item)
<div wire:key="$item->getRouteKey()">
  <livewire:item :key="$item->getRouteKey()" />
</div>
```

Note: I prefer the usage of route-keys, as they are more unique compared to the model id (slugs, hash-ids, etc.). It also helps against someone being able to count certain things, and you don't have to do `post-{$post->getKey()`.

You really need to be sure the keys are unique. Meaning when you push to an array for example (looping), make sure the key is unique as well. My mistake was pushing the same data, and thinking 'oh, Livewire will merge this for me', which isn't true unfortunately.

The `if` statement placing is also confusing. You have to surround that in a div, but sometimes it works fine without any. I think it happens because Livewire get's confused when you use multiple `if` statements like this:

```blade
@if ($foo) bar @endif
@if ($bar) foo @endif
@if ($e)
  @if ($e->type === 'exception')
  @endif
@endif
```

I perfectly understand your frustration, and I wished Livewire would had better official debug tools.

**Reply by chatisk** [reply]
Exactly. 

But didnt know that :key also should be used.  I thought it refers to alpinejs . Does it not ?

Or is it the short syntax for wire:key ?

**Reply by francoism90** [reply]
@chatisk It doesn't, see https://livewire.laravel.com/docs/troubleshooting#adding-wirekey

The `key` has no relation to Alpine, and it isn't a short syntax to `wire:key` as well (as far I understood the processing of Livewire components). The keys have different meaning and usage. This is what makes it so confusing to most users.

Also note:

```blade
<div>
@foreach($posts as $post)
    <div wire:key="{{ $post->id }}"> // 1
        ...
        <livewire:show-post :$post :key="$post->id" /> // 1
        ...
    </div>
@endforeach

@foreach($tags as $tag)
    <div wire:key="{{ $tag->id }}"> // 1
        ...
        <livewire:show-tag :$tag :key="$tag->id" /> // 1
        ...
    </div>
@endforeach
</div>
```
Means you have a conflict, both keys are the same, so Livewire doesn't know which part of the DOM should be tracked and updated.

That's why I prefer the usage of route-keys (or at least generated hashes), to make sure they can be tracked. Most will adopt something like `:wire:key="time()`, but this doesn't work. Yeah, it's unique, but Livewire doesn't know how to bind. The key shouldn't change, unless your view changes:

```blade
<div wire:key="container-{{ time() }}">
  <div wire:key="post-{{ time() }}">
  </div>
</div>
```

This may break, because it's not known by the Livewire state.

**Reply by chatisk** [reply]
So that means the doc is wrong ? I would put it as `wire:key="post-div-{{ $post->id }}"` and the other one `:key="'livewire-dummy'.$post->id"` if that's even achievable 😆

**Reply by francoism90** [reply]
Good question lol.

I think they should contain the same (key-)value, but I'm not sure.
You can still have different `wire:key`'s in the child component (e.g. `item.blade.php`). But when calling inside a loop, I believe the 'parent' should follow the pattern as stated in the doc.

The doc is correct, but it's missing information, like your example. :)
